chimera/precondition_evaluator.py

```diff
@@ -55,6 +56,44 @@ def __init__(self, sf_context, chimera_config_filepath, pge_config_filepath, set
                 file_name = '~/verdi/etc/settings.yaml'
             raise RuntimeError("Could not read settings file '{}': {}".format(file_name, e))
 
+    def repl_val_in_dict(self, d, val, job_params, root=None, optional_fields=None):
+        """
+        Recursive function to replace occurences of val in a dict with values from the job_params.
+        """
+
+        if root is None: root = []
+        if optional_fields is None: optional_fields = []
+        matched_keys = []
+        for k, v in d.items():
+            rt = copy.copy(root)
+            rt.append(k)
+            if isinstance(v, dict):
+                matched_keys.extend(self.repl_val_in_dict(v, val, job_params, rt, optional_fields))
+            if v == val:
+                jp_key = '.'.join(rt)
+                # use job_params with explicit dot notation
+                if jp_key in job_params:
+                    d[k] = job_params[jp_key]
+                    matched_keys.append(jp_key)
+                # maintain backwards-compatibility of using job_param values without dot notation
+                elif k in job_params:
+                    d[k] = job_params[k]
+                    matched_keys.append(k)
+                else:
+                    # check if optionalField; if so, set value to empty string
+                    if jp_key in optional_fields:
+                        logger.info("Explicit dot notation key {} is an optional field.".format(jp_key))
+                        logger.info("Setting {} to value to empty string.".format(k))
+                        d[k] = ""
+                    elif k in optional_fields:
+                        logger.info("Key {} is an optional field.".format(k))
+                        logger.info("Setting {} to value to empty string.".format(k))
```

Comment:
Looks like there is a typo. Maybe it should be as follows?
```suggestion
                        logger.info("Setting {} value to empty string.".format(k))
```